hotham/src/rendering/primitive.rs

```diff
@@ -139,14 +139,32 @@ impl Primitive {
     }
 
     /// Get a bounding sphere for the primitive, applying a transform
-    pub fn get_bounding_sphere(&self, transform: &Transform) -> Vector4<f32> {
-        let mut center = self.bounding_sphere.xyz();
-        let mut radius = self.bounding_sphere.w;
-
-        center += transform.translation;
-        radius *= transform.scale.max();
-
-        [center.x, center.y, center.z, radius].into()
+    pub fn get_bounding_sphere(&self, transform: &TransformMatrix) -> Vector4<f32> {
+        let center_in_local: Point3<_> = self.bounding_sphere.xyz().into();
+        let center_in_world =
+            Point3::<_>::from_homogeneous(transform.0 * center_in_local.to_homogeneous()).unwrap();
+
+        // The linear part contains the rotation and scale, we are interested in the scale.
+        let world_from_local_linear_part = transform.0.fixed_slice::<3, 3>(0, 0);
+
+        // The scale of the sphere is taken as the largest scale in any dimension.
+        // If the scale is non-uniform, there could be a tighter bounding sphere.
+        // If the scale is uniform, the quality of the bounding sphere will be unchanged.
+        let scale = world_from_local_linear_part
+            .column(0)
+            .magnitude_squared()
+            .max(world_from_local_linear_part.column(1).magnitude_squared())
+            .max(world_from_local_linear_part.column(2).magnitude_squared())
+            .sqrt();
```

Comment:
This can be replaced with an SDV if we want to handle the case when scale is applied non-uniformly in directions not aligned with local space.
```suggestion
            .svd_unordered(false, false)
            .singular_values
            .max();
```

Reply:
The worst case is when we have an extreme non-uniform scaling along the space diagonal. This gives a bound on the error of $\sqrt{3} \approx 1.732 $. We could use this as a fudge factor to guarantee correctness.